x-pack/plugin/ml/log-structure-finder/src/main/java/org/elasticsearch/xpack/ml/logstructure/JsonLogFileStructureFinder.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.ml.logstructure;
+
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.common.xcontent.DeprecationHandler;
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.xpack.ml.logstructure.TimestampFormatFinder.TimestampMatch;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.common.xcontent.json.JsonXContent.jsonXContent;
+
+/**
+ * Really ND-JSON.
+ */
+public class JsonLogFileStructureFinder extends AbstractStructuredLogFileStructureFinder implements LogFileStructureFinder {
+
+    private final List<String> sampleMessages;
+    private final LogFileStructure structure;
+
+    JsonLogFileStructureFinder(List<String> explanation, String sample, String charsetName, Boolean hasByteOrderMarker) throws IOException {
+
+        List<Map<String, ?>> sampleRecords = new ArrayList<>();
```

Comment:
It sees that the Finder classes' constructors are doing the bulk of the work here. I wonder if it would be best to avoid that by introducing a method that clearly communicates that processing is going to occur.

Reply:
I thought about this, but then the member variables cannot be `final`, and if you have one of these objects there's a potential doubt about whether the processing method has been called or not.

Reply:
Good point. The other idea would be to go with static factory methods. But I don't feel strongly about it.

Reply:
Good idea - I moved the major chunks of processing to static factory methods in https://github.com/elastic/elasticsearch/pull/32788/commits/baf863f94844b5b6ef5895437ab11cb6e997f3b1